Use @ arguments file for running java as part of native-image on Windows

quintesse opened the pull request:
A very simple solution to #2387 using the @ argument support added to `java` in version 9.

The other solution offered in the issue seems much more complex just for having it work on Java 8. If that's essential it seems more useful to backport the @ argument support to version 8 instead.

Btw, the code could be simplified a bit more by removing the length check and just always generate the @ file on windows. Not sure what would be preferable though.

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -95,6 +95,8 @@
     private static final String DEFAULT_GENERATOR_9PLUS_SUFFIX = "$JDK9Plus";
     private static final String CUSTOM_SYSTEM_CLASS_LOADER = NativeImageSystemClassLoader.class.getCanonicalName();
 
+    private static final int WINDOWS_MAX_COMMAND_LENGTH = 8192;
```
You could make this something closer to the [real limit](http://msdn.microsoft.com/en-us/library/ms682425%28VS.85%29.aspx) of 32768.

quintesse replied:
I wasn't actually sure what the real limit was. Google turned up 2047 for some Windows versions and 8191 for others (see https://support.microsoft.com/en-us/help/830473/command-prompt-cmd-exe-command-line-string-limitation). THe number of 32768 is also mentioned but for the size of the ENV vars.
If you are sure it's 32768 I'll be more than happy to change it.

quintesse replied:
Ah duh, I see you actually provided a link. Sorry about the noise, I'll change it ASAP :-)

quintesse replied:
Done.

quintesse commented:
Btw, the CI failed but not in something that seems to have to do anything with my PR?
It says the formatting of the file I changed is incorrect (but not in the part I changed) and to run `mx eclipseformat`. But when I run that it doesn't detect anything wrong (no changes are made). So I have no idea how to fix things.

quintesse commented:
About the OCA, I'm a Red Hat employee (tschotan@redhat.com), I sent in a signed copy of the agreement, but aren't we automatically covered already?

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
         int exitStatus = 1;
+        File argsFile = null;
         try {
+            if (OS.getCurrent() == OS.WINDOWS) {
```
Use `Platform.includedIn(Platform.WINDOWS.class)` instead of `OS.getCurrent() == OS.WINDOWS`. 
`Platform.includedIn(...)` is built-time constant. It ensures the code for special handling is not included in native-image images for Platforms != Windows.

quintesse replied:
Yes, I said the same in the PR description. Is that what you want me to do? Just remove the platform check and make this the default for all platforms?

quintesse replied:
Ok

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+                if (totalLength >= WINDOWS_MAX_COMMAND_LENGTH) {
+                    argsFile = File.createTempFile("native-image", "args");
+                    argsFile.deleteOnExit();
+                    Files.write(argsFile.toPath(), (Iterable<String>) command.stream().skip(1)::iterator);
```
Did you verify that this works if the arguments contain spaces (e.g. in paths)?
See https://docs.oracle.com/en/java/javase/14/docs/specs/man/java.html#java-command-line-argument-files

olpaw replied:
Looking at https://docs.oracle.com/en/java/javase/14/docs/specs/man/java.html#java-command-line-argument-files shows that some care is needed to ensure we do not run into problems with quoting and character escaping.

quintesse replied:
OK, arguments that have spaces are now quoted (not sure we need to check for other whitespace chars as well? Given the fact it's almost all generated instead of user-supplied?) and backspaces are duplicated correctly.

olpaw replied:
I'll this through all our gates and let you know if I find any quoting/escaping-related issues.

olpaw replied:
All internal CI gates are passing with a57da3f553aaec8ca353778bb5f24ccabff2fea1.

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+            if (OS.getCurrent() == OS.WINDOWS) {
+                int totalLength = command.stream().mapToInt(s -> s.length() + 1).sum();
+                if (totalLength >= WINDOWS_MAX_COMMAND_LENGTH) {
+                    argsFile = File.createTempFile("native-image", "args");
```
`Files#createTempDirectory` is preferred.

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
         try {
+            if (OS.getCurrent() == OS.WINDOWS) {
+                int totalLength = command.stream().mapToInt(s -> s.length() + 1).sum();
+                if (totalLength >= WINDOWS_MAX_COMMAND_LENGTH) {
```
If we make this the default for Java > 8 then we don't need to make this code Windows specific.

quintesse replied:
Yes, I said the same in the PR description. Is that what you want me to do? Just remove the platform check and make this the default for all platforms?

olpaw replied:
If we are confident after extensive testing that it causes no drawbacks then we should make that the default for Java > 8 (independent of the platform).

olpaw replied:
Please change the sources to use
```
        if (config.useJavaModules()) {
            /* For Java > 8 we use an argument file to pass the options to the builder */
```
instead of `if (Platform.includedIn(Platform.WINDOWS.class))` ... `if (totalLength >= WINDOWS_MAX_COMMAND_LENGTH) {`. Then I can run this PR in our CI to see if the argument file approach causes any trouble.

quintesse replied:
Done

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+                int totalLength = command.stream().mapToInt(s -> s.length() + 1).sum();
+                if (totalLength >= WINDOWS_MAX_COMMAND_LENGTH) {
+                    argsFile = File.createTempFile("native-image", "args");
+                    argsFile.deleteOnExit();
```
Do we really need `.deleteOnExit()` here if we already have `argsFile.delete();` in the finally-clause?

quintesse replied:
During testing I found out that hitting CTRL+C would not delete the file unless I also added the `.deleteOnExit()`. Now we could of course remove the `finally` block, because the process will never be long-lived, but it  seemed "untidy".

olpaw replied:
> Now we could of course remove the finally block, because the process will never be long-lived, but it seemed "untidy".

Would be fine for me.

quintesse replied:
Ok, done

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1203,11 +1204,30 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
         } catch (IOException | InterruptedException e) {
             throw showError(e.getMessage());
+        } finally {
+            if (argsFile != null) {
```
Use
```
`!useDebugAttach() && argsFile != null`
```
here.

quintesse replied:
I will, but could you explain that to me?

olpaw replied:
I you run `native-image` with `--debug-attach` you would eventually want to inspect the `argsFile` to know exactly what was passed passed to the native-image builder. See
https://github.com/oracle/graal/blob/75df2c1dc9ce10c175261f9d003eb983625a81b6/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java#L153
https://github.com/oracle/graal/blob/75df2c1dc9ce10c175261f9d003eb983625a81b6/substratevm/src/com.oracle.svm.driver/resources/HelpExtra.txt#L9

quintesse commented:
NB: I added an extra commit to fix a formatting problem in some other file that was tripping up the CI each time.

olpaw commented:
> NB: I added an extra commit to fix a formatting problem in some other file that was tripping up the CI each time.

Huh ... strange non of our internal CI gates has a problem with `substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/FirstObjectTable.java`. Do you have a link to the failing gate?

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1204,6 +1205,18 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+            if (config.useJavaModules()) {
+                /* For Java > 8 we use an argument file to pass the options to the builder */
+                Path argsFile = Files.createTempFile("native-image", "args");
+                argsFile.toFile().deleteOnExit();
```
Unconditionally deleting `argsFile` makes it harder to replay and debug failures. I would suggest only deleting it if `exitStatus` below is non-zero. In any case, no need to use `deleteOnExit` - just add a `finally` clause to the `try` statement.

quintesse replied:
@dougxc see the comments @olpaw made before. There was a try-finally deleting the file, but that doesn't delete the file when the user hits CTRL+C while deleteOnExit() does. So instead of having 2 ways of deleting it was decided to keep the one that works in more cases. For the debugging we an surround it with an `!useDebugAttach()` perhaps?

dougxc replied:
Ok, how about making a copy of the to-be-deleted-on-exit file if `exitStatus != 0`?

quintesse replied:
Wouldn't that leave a whole bunch of files around on the user's system? Remember that this is not a debug option or anything, it will be used on every execution.

dougxc replied:
I would suggest that the copy has a fixed name (e.g. `native-image-failure.args`) to avoid polluting the system. This file will deleted if it exists before running native-image only be left behind when native-image fails. In addition, all this can be activated only when `--verbose` is supplied. It's only in that scenario that someone may be interested in replaying the low-level native image command.

olpaw replied:
@quintesse are you going to address https://github.com/oracle/graal/pull/2443#discussion_r443523222 or should we close this PR? cc @dougxc

quintesse replied:
Done @olpaw @dougxc

olpaw commented:
Another problem is that this approach does not fix the problem for Windows native-images on Java 8. But its still an improvement over what we have now. cc @christianwimmer

quintesse commented:
> Do you have a link to the failing gate?

Sure, this is one of them: https://travis-ci.org/github/oracle/graal/builds/686028526

olpaw commented:
> > Do you have a link to the failing gate?
> 
> Sure @olpaw , this is one of them: https://travis-ci.org/github/oracle/graal/builds/686028526

@quintesse Thanks! @dougxc looks like `/home/travis/build/oracle/graal/../eclipse/eclipse` is different from what we use in our internal CI.

dougxc commented:
Indeed:
https://github.com/oracle/graal/blob/master/common.hocon#L111
https://github.com/oracle/graal/blob/master/.travis.yml#L73

@quintesse can you please add this patch to this PR to see if it fixes the problem:
```
diff --git a/.travis.yml b/.travis.yml
index 24bc84f0c4f..bc55122bbfb 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -70,7 +70,7 @@ install:
         pip install pylint==1.1.0

         export ECLIPSE_TAR=${TRAVIS_BUILD_DIR}/../eclipse.tar.gz
-        wget https://archive.eclipse.org/eclipse/downloads/drops4/R-4.5.2-201602121500/eclipse-SDK-4.5.2-linux-gtk-x86_64.tar.gz -O ${ECLIPSE_TAR}
+        wget https://archive.eclipse.org/eclipse/downloads/drops4/R-4.14-201912100610/eclipse-SDK-4.14-linux-gtk-x86_64.tar.gz -O ${ECLIPSE_TAR}
         tar -C ${TRAVIS_BUILD_DIR}/.. -xzf ${ECLIPSE_TAR}
         export ECLIPSE_EXE=${TRAVIS_BUILD_DIR}/../eclipse/eclipse
```

quintesse commented:
> @quintesse can you please add this patch to this PR to see if it fixes the problem:

Sure @dougxc , done.

quintesse commented:
Oops, forgot I had to first remove my previous fix of course. Repushed.

quintesse commented:
Seems that fixed it @dougxc

jaikiran commented:
One of the users of Quarkus reported this issue recently https://github.com/quarkusio/quarkus/issues/10074. 

It would be good to have this PR merged and available in next release of Graal.

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1255,14 +1257,41 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+                /* Create a copy of the argument file for failure analysis */
+                Path tmpDir = Paths.get(System.getProperty("java.io.tmpdir"));
+                Path argsFileCopy = tmpDir.resolve("native-image-failure.args");
+                Files.copy(argsFile, argsFileCopy, StandardCopyOption.REPLACE_EXISTING);
```
How is the user supposed to find this copy? The location of the temp dir is not obvious on all OSes. For example, on my Mac it's `/var/folders/wr/4yqmgc5978z05qsv8j4sxjlr0000gn/T/` which is pretty hard to guess ;-) I would suggest putting the copy in the current working directory. In any case, you should issue a println saying where it is.

olpaw replied:
How about using https://github.com/oracle/graal/blob/146a5548cf95c4803c10103f1e67a0e59dd93ce7/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L695 ?

quintesse replied:
Implemented @dougxc 's suggestion. But personally I'd just a) not delete the temp file and b) print its location (only when `isVerbose()` is true). It's the simplest implementation, it uses temp files in a location meant for temp files and it's still easy to find (once we print the path).

dougxc replied:
I still prefer the option of retaining the original temp file so that the command line can simply be copied and pasted for reproduction. You could achieve this with a "conditional" shutdown hook:
```
        Path argsFile = Files.createTempFile("native-image", "args");
        Path[] argsFileBox = {argsFile};
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
                if (argsFileBox[0] != null) {
                    argsFileBox[0].toFile().delete();
                }
            }
        });

        ...

        if (exitStatus != 0 && isVerbose() && argsFile != null) {
           argsFileBox[0] = null;
        }
```

quintesse replied:
Like this?

dougxc replied:
Yes.

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1255,14 +1257,50 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
             exitStatus = p.waitFor();
+            if (exitStatus != 0 && isVerbose() && argsFile != null) {
+                argsFileBox[0] = null;
+                showMessage("Argument file: " + argsFile);
```
You can remove the `showMessage` line now.

quintesse replied:
But then how will people find the file as you mentioned before?

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1255,14 +1257,50 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
 
         int exitStatus = 1;
         try {
+            Path argsFile = null;
```
Move declaration of `argsFile` into the scope of `if (config.useJavaModules()) {`.

quintesse replied:
ok

dougxc commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:
```diff
@@ -1255,14 +1257,50 @@ protected int buildImage(List<String> javaArgs, LinkedHashSet<Path> bcp, LinkedH
+                        }
+                    }
+                });                
+                argsFile = Files.createTempFile("native-image", "args");
```
These 2 lines should now be deleted.

quintesse replied:
oops, of course

quintesse commented:
Applied latest requested changes

olpaw commented:
@quintesse you changed the PR while it was already merging. Therefore this PR did not close automatically. But your changes are on master: https://github.com/oracle/graal/commit/9a9039f3f3b6ff845dab153a6746155bdb7f4e9a

quintesse commented:
> you changed the PR while it was already merging

Hi, I did so because there was a failure in one of the integration tests telling me there was a formatting error. So I thought I had to fix that first. 

Thanks for merging.
